File: CSharp/SpaceVIL/SpaceVIL/Engine/Shader.cs
using System;
using System.Text;
using System.IO;
using Glfw3;

#if OS_LINUX
using static GL.LGL.OpenLGL;
#elif OS_WNDOWS
using static GL.WGL.OpenWGL;
#else
using static GL.WGL.OpenWGL;
#endif

namespace SpaceVIL
{
    internal class Shader// : GL.WGL.OpenWGL
    {
        const uint Vertex = 0x8B31;
        const uint Fragment = 0x8B30;
        const uint Geometry = 0x8DD9;
        const uint Compute = 0x91B9;


        private uint _program_id;
        internal uint GetProgramID()
        {
            return _program_id;
        }

        private uint _shader_vertex;
        private String _code_vertex;

        private uint _shader_fragment;
        private String _code_fragment;

        private uint _shader_geometry;
        private String _code_geometry;

        private uint _shader_compute;
        private String _code_compute;

        public String GetCode(uint shader_type)
        {
            switch (shader_type)
            {
                case Vertex:
                    return _code_vertex;

                case Fragment:
                    return _code_fragment;

                case Geometry:
                    return _code_geometry;

                case Compute:
                    return _code_compute;

                default:
                    break;
            }
            return String.Empty;
        }
        public void SetShaderPath(uint shader_type, String path)
        {
            String code = ReadSource(path);

            switch (shader_type)
            {
                case Vertex:
                    _code_vertex = code.ToString();
                    break;

                case Fragment:
                    _code_fragment = code.ToString();
                    break;

                case Geometry:
                    _code_geometry = code.ToString();
                    break;

                case Compute:
                    _code_compute = code.ToString();
                    break;

                default:
                    break;
            }
        }

        internal Shader()
        {
            InitDefaults();
        }
        internal Shader(String vertex_code, String fragmend_code)
        {
            _code_vertex = vertex_code;
            _code_fragment = fragmend_code;

            //defaults
            _code_geometry = String.Empty;
            _code_compute = String.Empty;
        }
        internal Shader(Stream vertex_code, Stream fragmend_code)
        {
            _code_vertex = ReadSource(vertex_code);
            _code_fragment = ReadSource(fragmend_code);

            //defaults
            _code_geometry = String.Empty;
            _code_compute = String.Empty;
        }
        private void InitDefaults()
        {
            _code_vertex = String.Empty;
            _code_fragment = String.Empty;
            _code_geometry = String.Empty;
            _code_compute = String.Empty;
        }
        private String ReadSource(Stream source)
        {
            StringBuilder code = new StringBuilder();
            try
            {
                using (StreamReader sr = new StreamReader(source))
                {
                    while (!sr.EndOfStream)
                        code.Append(sr.ReadLine() + "\n");
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("The stream could not be read:");
                Console.WriteLine(e.Message);
            }
            return code.ToString();
        }
        private String ReadSource(String source)
        {
            StringBuilder code = new StringBuilder();
            try
            {
                using (StreamReader sr = new StreamReader(source))
                {
                    while (!sr.EndOfStream)
                        code.Append(sr.ReadLine() + "\n");
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("The file could not be read:");
                Console.WriteLine(e.Message);
            }
            return code.ToString();
        }

        internal void Compile()
        {
            _program_id = glCreateProgram();
            //compiling
            if (_code_vertex != String.Empty)
            {
                _shader_vertex = CompileShader(Vertex, _code_vertex);
                glAttachShader(_program_id, _shader_vertex);
            }

            if (_code_fragment != String.Empty)
            {
                _shader_fragment = CompileShader(Fragment, _code_fragment);
                glAttachShader(_program_id, _shader_fragment);
            }

            if (_code_geometry != String.Empty)
            {
                _shader_geometry = CompileShader(Geometry, _code_geometry);
                glAttachShader(_program_id, _shader_geometry);
            }

            if (_code_compute != String.Empty)
            {
                _shader_compute = CompileShader(Compute, _code_compute);
                glAttachShader(_program_id, _shader_compute);
            }
            //linking
            glLinkProgram(_program_id);
            
        }
        private uint CompileShader(uint shader_type, String code)
        {
            uint shader_id = glCreateShader((uint)shader_type);
            glShaderSource(shader_id, 1, new[] { code }, new[] { code.Length });
            glCompileShader(shader_id);
            return shader_id;
        }

        internal void UseShader()
        {
            glUseProgram(_program_id);
        }

        internal void DeleteShader()
        {
            DetachShadersCode();
            DeleteShadersID();
            glDeleteProgram(_program_id);
        }
        private void DetachShadersCode()
        {
            glDetachShader(_program_id, _shader_vertex);
            glDetachShader(_program_id, _shader_fragment);
            glDetachShader(_program_id, _shader_geometry);
            glDetachShader(_program_id, _shader_compute);
        }
        private void DeleteShadersID()
        {
            glDeleteShader(_shader_vertex);
            glDeleteShader(_shader_fragment);
            glDeleteShader(_shader_geometry);
            glDeleteShader(_shader_compute);
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/Shader.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.internal.Wrapper.OpenGLWrapper;

import java.io.FileReader;
import java.io.BufferedReader;

final class Shader {
    private OpenGLWrapper gl = null;

    private final int VERTEX = 0x8B31;
    private final int FRAGMENT = 0x8B30;
    private final int GEOMETRY = 0x8DD9;
    private final int COMPUTE = 0x91B9;


    private int _programId;

    long getProgramID() {
        return _programId;
    }

    private int _shaderVertex;
    private String _codeVertex;

    private int _shaderFragment;
    private String _codeFragment;

    private int _shaderGeometry;
    private String _codeGeometry;

    private int _shaderCompute;
    private String _codeCompute;

    String getCode(int shaderType) {
        switch (shaderType) {
        case VERTEX:
            return _codeVertex;

        case FRAGMENT:
            return _codeFragment;

        case GEOMETRY:
            return _codeGeometry;

        case COMPUTE:
            return _codeCompute;

        default:
            break;
        }
        return "";
    }

    void setShaderPath(int shaderType, String path) {
        String code = readSource(path);

        switch (shaderType) {
        case VERTEX:
            _codeVertex = code.toString();
            break;

        case FRAGMENT:
            _codeFragment = code.toString();
            break;

        case GEOMETRY:
            _codeGeometry = code.toString();
            break;

        case COMPUTE:
            _codeCompute = code.toString();
            break;

        default:
            break;
        }
    }

    Shader() {
        gl = OpenGLWrapper.get();
        initDefaults();
    }

    Shader(String vertexCode, String fragmendCode) {
        gl = OpenGLWrapper.get();
        _codeVertex = vertexCode;
        _codeFragment = fragmendCode;

        // defaults
        _codeGeometry = "";
        _codeCompute = "";
    }

    Shader(BufferedReader vertexCode, BufferedReader fragmendCode) {
        gl = OpenGLWrapper.get();
        _codeVertex = readSource(vertexCode);
        _codeFragment = readSource(fragmendCode);

        // defaults
        _codeGeometry = "";
        _codeCompute = "";
    }

    private void initDefaults() {
        _codeVertex = "";
        _codeFragment = "";
        _codeGeometry = "";
        _codeCompute = "";
    }

    private String readSource(BufferedReader source) {
        StringBuilder code = new StringBuilder();
        String line = null;
        try {
            while ((line = source.readLine()) != null) {
                code.append(line);
                code.append('\n');
            }
        } catch (Exception e) {
            System.out.println("The stream could not be read:");
            System.out.println(e.getMessage());
        }
        return code.toString();
    }

    private String readSource(String source) {
        StringBuilder code = new StringBuilder();
        String line = null;
        try {
            BufferedReader reader = new BufferedReader(new FileReader(source));
            while ((line = reader.readLine()) != null) {
                code.append(line);
                code.append('\n');
            }
            reader.close();
        }

        catch (Exception e) {
            System.out.println("The file could not be read:");
            System.out.println(e.getMessage());
        }
        return code.toString();
    }

    void compile() {
        _programId = gl.CreateProgram();
        // compiling
        if (!_codeVertex.equals("")) {
            _shaderVertex = compileShader(VERTEX, _codeVertex);
            gl.AttachShader(_programId, _shaderVertex);
        }

        if (!_codeFragment.equals("")) {
            _shaderFragment = compileShader(FRAGMENT, _codeFragment);
            gl.AttachShader(_programId, _shaderFragment);
        }

        if (!_codeGeometry.equals("")) {
            _shaderGeometry = compileShader(GEOMETRY, _codeGeometry);
            gl.AttachShader(_programId, _shaderGeometry);
        }

        if (!_codeCompute.equals("")) {
            _shaderCompute = compileShader(COMPUTE, _codeCompute);
            gl.AttachShader(_programId, _shaderCompute);
        }
        // linking
        gl.LinkProgram(_programId);

    }

    private int compileShader(int shaderType, String code) {
        int shaderId = gl.CreateShader(shaderType);
        gl.ShaderSource(shaderId, code);
        gl.CompileShader(shaderId);
        return shaderId;
    }

    void useShader() {
        gl.UseProgram(_programId);
    }

    void deleteShader() {
        detachShaders();
        deleteShadersID();
        gl.DeleteProgram(_programId);
    }

    private void detachShaders() {
        gl.DetachShader(_programId, _shaderVertex);
        gl.DetachShader(_programId, _shaderFragment);
        gl.DetachShader(_programId, _shaderGeometry);
        gl.DetachShader(_programId, _shaderCompute);
    }

    private void deleteShadersID() {
        gl.DeleteShader(_shaderVertex);
        gl.DeleteShader(_shaderFragment);
        gl.DeleteShader(_shaderGeometry);
        gl.DeleteShader(_shaderCompute);
    }
}